crates/core/tedge/src/cli/connect/command.rs

```diff
@@ -567,3 +573,136 @@ fn get_common_mosquitto_config_file_path(
         .join(TEDGE_BRIDGE_CONF_DIR_PATH)
         .join(&common_mosquitto_config.config_file)
 }
+
+// To confirm the connected c8y tenant is the one that user configured.
+fn check_connected_c8y_tenant_as_configured(configured_url: &str, port: u16) {
+    match get_connected_c8y_url(port) {
+        Ok(url) if url == configured_url => {}
+        Ok(url) => println!(
+            "Warning: Connecting to {}, but configured URL is {}.\n",
+            url, configured_url
+        ),
+        Err(_) => println!("Failed to get the connected tenant URL from Cumulocity.\n"),
+    }
+}
+
+fn get_connected_c8y_url(port: u16) -> Result<String, ConnectError> {
+    const C8Y_TOPIC_BUILTIN_JWT_TOKEN_UPSTREAM: &str = "c8y/s/uat";
+    const C8Y_TOPIC_BUILTIN_JWT_TOKEN_DOWNSTREAM: &str = "c8y/s/dat";
+    const CLIENT_ID: &str = "get_jwt_token_c8y";
+
+    let mut options = MqttOptions::new(CLIENT_ID, DEFAULT_HOST, port);
+    options.set_keep_alive(RESPONSE_TIMEOUT);
+
+    let (mut client, mut connection) = rumqttc::Client::new(options, 10);
+    let mut acknowledged = false;
+
+    client.subscribe(C8Y_TOPIC_BUILTIN_JWT_TOKEN_DOWNSTREAM, AtLeastOnce)?;
+
+    for event in connection.iter() {
+        match event {
+            Ok(Event::Incoming(Packet::SubAck(_))) => {
+                // We are ready to get the response, hence send the request
+                client.publish(C8Y_TOPIC_BUILTIN_JWT_TOKEN_UPSTREAM, AtLeastOnce, false, "")?;
+            }
+            Ok(Event::Incoming(Packet::PubAck(_))) => {
+                // The request has been sent
+                acknowledged = true;
+            }
+            Ok(Event::Incoming(Packet::Publish(response))) => {
+                // We got a response
+                let token = String::from_utf8(response.payload.to_vec()).unwrap();
+                let connected_url = decode_jwt_token(token.as_str())?;
+                return Ok(connected_url);
+            }
+            Ok(Event::Outgoing(Outgoing::PingReq)) => {
+                // No messages have been received for a while
+                println!("Local MQTT publish has timed out.");
+                break;
+            }
+            Ok(Event::Incoming(Incoming::Disconnect)) => {
+                eprintln!("ERROR: Disconnected");
+                break;
+            }
+            Err(err) => {
+                eprintln!("ERROR: {:?}", err);
+                break;
+            }
+            _ => {}
+        }
+    }
+
+    if acknowledged {
+        // The request has been sent but without a response
+        println!("\nThe request has been sent, however, no response.");
+        Err(ConnectError::TimeoutElapsedError)
+    } else {
+        // The request has not even been sent
+        println!("\nMake sure mosquitto is running.");
+        Err(ConnectError::TimeoutElapsedError)
+    }
+}
```

Comment:
This function seems to be doing a little more than just `get_connected_c8y_url`, I suppose it would be good to extract connection and the loop from here to other functions

Reply:
Cannot we reuse some of the code already existing in the mapper for JWT retrieval?

Reply:
In the mapper, the code to process JWT tokens is really light since one just uses them without any decoding.

But it would be good to extract the JWT token code out of this file into a jwt-token module under `cli/connect`.